pkg/command/plugin.go

```diff
@@ -540,3 +565,227 @@ func displayInstalledAndMissingListView(installedStandalonePlugins []cli.PluginI
 func getTarget() configtypes.Target {
 	return configtypes.StringToTarget(strings.ToLower(targetStr))
 }
+
+// ====================================
+// Shell completion functions
+// ====================================
+func completeInstalledPlugins(_ *cobra.Command, args []string, _ string) ([]string, cobra.ShellCompDirective) {
+	installedPlugins, err := pluginsupplier.GetInstalledPlugins()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	var comps []string
+	target := getTarget()
+	if len(args) == 0 {
+		// Complete all plugin names as long as the target matches and let the shell filter
+		for i := range installedPlugins {
+			if target == configtypes.TargetUnknown || target == installedPlugins[i].Target {
+				// Make sure the name of the plugin is part of the description so that
+				// zsh does not lump many plugins that have the same description
+				comps = append(comps, fmt.Sprintf("%[1]s\tTarget: %[2]s for %[1]s", installedPlugins[i].Name, installedPlugins[i].Target))
+			}
+		}
+	}
+
+	comps = completionMergeSimilarPlugins(comps)
+
+	return comps, cobra.ShellCompDirectiveNoFileComp
+}
+
+func completeAllPlugins(_ *cobra.Command, args []string, _ string) ([]string, cobra.ShellCompDirective) {
+	if len(args) > 0 {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+	return completionAllPlugins(), cobra.ShellCompDirectiveNoFileComp
+}
+
+func completePluginVersions(_ *cobra.Command, args []string, _ string) ([]string, cobra.ShellCompDirective) {
+	if len(args) == 0 {
+		// We can't complete the version if we don't have a plugin name
+		comps := cobra.AppendActiveHelp(nil, "You must first specify a plugin name to be able to complete its version")
+		return comps, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	criteria := &discovery.PluginDiscoveryCriteria{
+		Name:   args[0],
+		Target: configtypes.StringToTarget(targetStr),
+	}
+
+	plugins, err := pluginmanager.DiscoverStandalonePlugins(
+		discovery.WithPluginDiscoveryCriteria(criteria),
+		discovery.WithUseLocalCacheOnly())
+
+	if err != nil || len(plugins) == 0 {
+		return nil, cobra.ShellCompDirectiveNoFileComp
```

Comment:
wondering: whether this also a situation that user has done something wrong (provided an "incorrect" plugin name) and if we could do something similar to L628 to inform the user

Reply:
That is a good suggestion.  I will push an update that will improve this as follows:

```
$ tz plugin install clusterssss -v <TAB>
Unable to find plugin 'clusterssss'

$ tz plugin install policy -t k8s -v <TAB>
Unable to find plugin 'policy' for target 'k8s'
```

Note that we can also do a check for when a target is invalid after the `--target` flag.  I didn't do it in this case because I prefer doing it every time we have an invalid target, not just for the `--version` flag completion.  I can do that in another PR.